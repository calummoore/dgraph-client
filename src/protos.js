import grpc from 'grpc';

const {
  graphp,
  facetsp,
  taskp,
  workerp,
} = grpc.load(__dirname + '/../protos/workerp/payload.proto');
const { typesp } = grpc.load(__dirname + "/../protos/typesp/types.proto");

const protos = {
  graphp,
  facetsp,
  typesp,
  taskp,
  workerp,
};

module.exports = protos;
