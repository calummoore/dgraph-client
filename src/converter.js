/* @flow */
import protos from './protos';
import wkx from 'wkx';

const {
  graphp,
  typesp
} = protos;

export function setPropertyValue(nquad, value) {
  switch (typeof value) {
    case 'string':
      nquad.objectType = typesp.Posting.ValType.STRING;
      nquad.objectValue = new graphp.Value().set('str_val', value);
      break;
    case 'number':
      nquad.objectType = typesp.Posting.ValType.FLOAT;
      nquad.objectValue = new graphp.Value().set('double_val', value);
      break;
    case 'boolean':
      nquad.objectType = typesp.Posting.ValType.BOOL;
      nquad.objectValue = new graphp.Value().set('bool_val', value);
      break;
    case 'object':
      // GEO JSON
      if((value.type === 'Point' || value.type === 'Polygon')
        && value.coordinates.constructor === Array){
        nquad.objectType = typesp.Posting.ValType.GEO;
        let wkb = wkx.Geometry.parseGeoJSON(value).toWkb();
        nquad.objectValue = new graphp.Value().set('geo_val', wkb);
      } else if (value instanceof Date){
        nquad.objectType = typesp.Posting.ValType.DATETIME;
        nquad.objectValue = new graphp.Value().set('datetime_val', value.toISOString())
      }
      break;
  }
}

export function toSetMutation(object, map = []){
  let mutation = new graphp.Mutation();

  function itterObj(obj, map, counter = 0) {
    if(map.includes(obj)){
      return obj._uid_ || obj.__tmpId;
    }

    let id = obj._uid_ || '_:' + (obj.__tmpId = 'tmp' + counter);
    map.push(obj);

    for (let key in obj) {
      if (!obj.hasOwnProperty(key) || key === '_uid_' || key === '__tmpId')
        continue;

      let nquad = new graphp.NQuad();
      nquad.subject = id;
      nquad.predicate = key;

      let value = obj[key];
      setPropertyValue(nquad, value);

      if (nquad.objectValue === null && typeof value === 'object') {
        nquad.objectId = itterObj(value, map, counter + 1);
      }

      mutation.set.push(nquad);
    }

    return id;
  }
  itterObj(object, map);

  return mutation;
}

export function getPropertyValue(prop) {
  switch (prop.value.val){
    case 'default_val':
    case 'str_val':
    case 'bool_val':
    case 'double_val':
    case 'int_val':
      return prop.value[prop.value.val];
    case 'geo_val':
      return wkx.Geometry.parse(prop.value.geo_val).toGeoJSON();
    case 'date_val':
    case 'datetime_val':
      return Date.parse(prop.value[prop.value.val]);
    case 'bytes_val':
    case 'password_val':
    default:
      return undefined;
  }
}

export function nodeToObject(node) {
  let obj = {};
  if(node.uid && node.uid > 0){
    obj._uid_ = node.uid;
  }
  if(node.xid){
    obj._xid_ = node.xid;
  }
  if (node.children) {
    for (let child of node.children) {
      obj[child.attribute] = nodeToObject(child);
    }
  }
  if (node.properties) {
    for (let prop of node.properties) {
      obj[prop.prop] = getPropertyValue(prop);
    }
  }
  return obj;
}