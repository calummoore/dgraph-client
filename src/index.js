// @flow

import grpc from 'grpc'
import protos from './protos'
import { nodeToObject, toSetMutation } from './converter'

/**
 * A dgraph gRPC client
 */
class DgraphClient {
  dgraph: protos.Dgraph

  /**
   * create new client
   * @param {string} url
   * @param credentials (optional)
   */
  constructor (url: string, credentials: any = undefined) {
    if (!credentials) {
      credentials = grpc.credentials.createInsecure()
    }

    this.dgraph = new protos.Dgraph(url, credentials)
  }

  /**
   * Run a DgraphClient.Request against dgraph and get raw response
   * @param request
   * @returns {Promise}
   */
  run (request: any) {
    return new Promise((resolve, reject) => {
      this.dgraph.run(request, (err, response) => {
        if (err) {
          reject(err)
        } else {
          resolve(response)
        }
      })
    })
  }

  /**
   * Run a query against dgraph
   * @param query
   */
  query (query: string): Promise<Object> {
    let request = new protos.Request()
    request.query = query
    return this.run(request).then(response => nodeToObject(response.n[0]))
  }

  /**
   * Set mutation against dgraph
   * @param object
   * @returns {Promise.<*>}
   */
  set<T: Object> (object: T): Promise<T & {id: string}> {
    let request = new protos.Request()

    let map = []
    request.mutation = toSetMutation(object, map)

    return this.run(request)
      .then(response => {
        for (let obj of map) {
          if (obj.__tmpId) {
            obj['_uid_'] = response.AssignedUids[obj.__tmpId]
            // $FlowFixMe
            delete obj.__tmpId
          }
        }

        return object
      })
  }
}

export default DgraphClient
