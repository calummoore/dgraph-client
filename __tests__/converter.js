/**
 * Copyright (c) 2017 VLEK Technology UG (haftungsbeschrÃ¤nkt). All rights reserved.
 *
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * proprietary and confidential.
 */
/* eslint-env jest */

import {setPropertyValue, toSetMutation} from '../src/converter'
import protos from '../src/protos'

describe('setPropertyValue', () => {
  it('can set a string', () => {
    let nquad = new protos.NQuad()
    setPropertyValue(nquad, 'string test')

    expect(nquad).toMatchSnapshot()
  })

  it('can set a double', () => {
    let nquad = new protos.NQuad()
    setPropertyValue(nquad, 1.253)

    expect(nquad).toMatchSnapshot()
  })

  it('can set a bool', () => {
    let nquad = new protos.NQuad()
    setPropertyValue(nquad, true)

    expect(nquad).toMatchSnapshot()
  })

  it('can set a geopoint', () => {
    let nquad = new protos.NQuad()
    setPropertyValue(nquad, {type: 'Point', coordinates: [13.725, 52.145]})

    expect(nquad).toMatchSnapshot()
  })
})

describe('toSetMutation', () => {
  it('create nquads from object', () => {
    let obj = {name: 'Paul', size: 1.85}
    let mutations = toSetMutation(obj, [])

    expect(mutations).toMatchSnapshot()
  })

  it('uses _uid_ as subject', () => {
    let obj = {_uid_: '0xd8450668580bf30f', name2: 'Paul 2'}
    let mutations = toSetMutation(obj, [])

    expect(mutations).toMatchSnapshot()
  })

  it('create nquads for nested objects', () => {
    let obj = {name: 'Paul', friend: {name: 'Max'}}
    let mutations = toSetMutation(obj, [])

    expect(mutations).toMatchSnapshot()
  })
})
