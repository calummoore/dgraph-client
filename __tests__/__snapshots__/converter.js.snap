// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`setPropertyValue can set a bool 1`] = `
 {
  "facets": Array [],
  "label": "",
  "lang": "",
  "objectId": "",
  "objectType": 4,
  "objectValue":  {
    "bool_val": true,
    "bytes_val": ByteBuffer {
      "buffer": Buffer [],
      "limit": 0,
      "littleEndian": false,
      "markedOffset": -1,
      "noAssert": false,
      "offset": 0,
    },
    "date_val": ByteBuffer {
      "buffer": Buffer [],
      "limit": 0,
      "littleEndian": false,
      "markedOffset": -1,
      "noAssert": false,
      "offset": 0,
    },
    "datetime_val": ByteBuffer {
      "buffer": Buffer [],
      "limit": 0,
      "littleEndian": false,
      "markedOffset": -1,
      "noAssert": false,
      "offset": 0,
    },
    "default_val": "",
    "double_val": 0,
    "geo_val": ByteBuffer {
      "buffer": Buffer [],
      "limit": 0,
      "littleEndian": false,
      "markedOffset": -1,
      "noAssert": false,
      "offset": 0,
    },
    "int_val": Long {
      "high": 0,
      "low": 0,
      "unsigned": false,
    },
    "password_val": "",
    "str_val": "",
    "uid_val": Long {
      "high": 0,
      "low": 0,
      "unsigned": true,
    },
    "val": "bool_val",
  },
  "predicate": "",
  "subject": "",
}
`;

exports[`setPropertyValue can set a double 1`] = `
 {
  "facets": Array [],
  "label": "",
  "lang": "",
  "objectId": "",
  "objectType": 3,
  "objectValue":  {
    "bool_val": false,
    "bytes_val": ByteBuffer {
      "buffer": Buffer [],
      "limit": 0,
      "littleEndian": false,
      "markedOffset": -1,
      "noAssert": false,
      "offset": 0,
    },
    "date_val": ByteBuffer {
      "buffer": Buffer [],
      "limit": 0,
      "littleEndian": false,
      "markedOffset": -1,
      "noAssert": false,
      "offset": 0,
    },
    "datetime_val": ByteBuffer {
      "buffer": Buffer [],
      "limit": 0,
      "littleEndian": false,
      "markedOffset": -1,
      "noAssert": false,
      "offset": 0,
    },
    "default_val": "",
    "double_val": 1.253,
    "geo_val": ByteBuffer {
      "buffer": Buffer [],
      "limit": 0,
      "littleEndian": false,
      "markedOffset": -1,
      "noAssert": false,
      "offset": 0,
    },
    "int_val": Long {
      "high": 0,
      "low": 0,
      "unsigned": false,
    },
    "password_val": "",
    "str_val": "",
    "uid_val": Long {
      "high": 0,
      "low": 0,
      "unsigned": true,
    },
    "val": "double_val",
  },
  "predicate": "",
  "subject": "",
}
`;

exports[`setPropertyValue can set a geopoint 1`] = `
 {
  "facets": Array [],
  "label": "",
  "lang": "",
  "objectId": "",
  "objectType": 7,
  "objectValue":  {
    "bool_val": false,
    "bytes_val": ByteBuffer {
      "buffer": Buffer [],
      "limit": 0,
      "littleEndian": false,
      "markedOffset": -1,
      "noAssert": false,
      "offset": 0,
    },
    "date_val": ByteBuffer {
      "buffer": Buffer [],
      "limit": 0,
      "littleEndian": false,
      "markedOffset": -1,
      "noAssert": false,
      "offset": 0,
    },
    "datetime_val": ByteBuffer {
      "buffer": Buffer [],
      "limit": 0,
      "littleEndian": false,
      "markedOffset": -1,
      "noAssert": false,
      "offset": 0,
    },
    "default_val": "",
    "double_val": 0,
    "geo_val": ByteBuffer {
      "buffer": Object {
        "data": Array [
          1,
          1,
          0,
          0,
          0,
          51,
          51,
          51,
          51,
          51,
          115,
          43,
          64,
          195,
          245,
          40,
          92,
          143,
          18,
          74,
          64,
        ],
        "type": "Buffer",
      },
      "limit": 21,
      "littleEndian": false,
      "markedOffset": -1,
      "noAssert": false,
      "offset": 0,
    },
    "int_val": Long {
      "high": 0,
      "low": 0,
      "unsigned": false,
    },
    "password_val": "",
    "str_val": "",
    "uid_val": Long {
      "high": 0,
      "low": 0,
      "unsigned": true,
    },
    "val": "geo_val",
  },
  "predicate": "",
  "subject": "",
}
`;

exports[`setPropertyValue can set a string 1`] = `
 {
  "facets": Array [],
  "label": "",
  "lang": "",
  "objectId": "",
  "objectType": 10,
  "objectValue":  {
    "bool_val": false,
    "bytes_val": ByteBuffer {
      "buffer": Buffer [],
      "limit": 0,
      "littleEndian": false,
      "markedOffset": -1,
      "noAssert": false,
      "offset": 0,
    },
    "date_val": ByteBuffer {
      "buffer": Buffer [],
      "limit": 0,
      "littleEndian": false,
      "markedOffset": -1,
      "noAssert": false,
      "offset": 0,
    },
    "datetime_val": ByteBuffer {
      "buffer": Buffer [],
      "limit": 0,
      "littleEndian": false,
      "markedOffset": -1,
      "noAssert": false,
      "offset": 0,
    },
    "default_val": "",
    "double_val": 0,
    "geo_val": ByteBuffer {
      "buffer": Buffer [],
      "limit": 0,
      "littleEndian": false,
      "markedOffset": -1,
      "noAssert": false,
      "offset": 0,
    },
    "int_val": Long {
      "high": 0,
      "low": 0,
      "unsigned": false,
    },
    "password_val": "",
    "str_val": "string test",
    "uid_val": Long {
      "high": 0,
      "low": 0,
      "unsigned": true,
    },
    "val": "str_val",
  },
  "predicate": "",
  "subject": "",
}
`;

exports[`toSetMutation create nquads for nested objects 1`] = `
 {
  "del": Array [],
  "schema": Array [],
  "set": Array [
     {
      "facets": Array [],
      "label": "",
      "lang": "",
      "objectId": "",
      "objectType": 10,
      "objectValue":  {
        "bool_val": false,
        "bytes_val": ByteBuffer {
          "buffer": Buffer [],
          "limit": 0,
          "littleEndian": false,
          "markedOffset": -1,
          "noAssert": false,
          "offset": 0,
        },
        "date_val": ByteBuffer {
          "buffer": Buffer [],
          "limit": 0,
          "littleEndian": false,
          "markedOffset": -1,
          "noAssert": false,
          "offset": 0,
        },
        "datetime_val": ByteBuffer {
          "buffer": Buffer [],
          "limit": 0,
          "littleEndian": false,
          "markedOffset": -1,
          "noAssert": false,
          "offset": 0,
        },
        "default_val": "",
        "double_val": 0,
        "geo_val": ByteBuffer {
          "buffer": Buffer [],
          "limit": 0,
          "littleEndian": false,
          "markedOffset": -1,
          "noAssert": false,
          "offset": 0,
        },
        "int_val": Long {
          "high": 0,
          "low": 0,
          "unsigned": false,
        },
        "password_val": "",
        "str_val": "Paul",
        "uid_val": Long {
          "high": 0,
          "low": 0,
          "unsigned": true,
        },
        "val": "str_val",
      },
      "predicate": "name",
      "subject": "_:tmp0",
    },
     {
      "facets": Array [],
      "label": "",
      "lang": "",
      "objectId": "",
      "objectType": 10,
      "objectValue":  {
        "bool_val": false,
        "bytes_val": ByteBuffer {
          "buffer": Buffer [],
          "limit": 0,
          "littleEndian": false,
          "markedOffset": -1,
          "noAssert": false,
          "offset": 0,
        },
        "date_val": ByteBuffer {
          "buffer": Buffer [],
          "limit": 0,
          "littleEndian": false,
          "markedOffset": -1,
          "noAssert": false,
          "offset": 0,
        },
        "datetime_val": ByteBuffer {
          "buffer": Buffer [],
          "limit": 0,
          "littleEndian": false,
          "markedOffset": -1,
          "noAssert": false,
          "offset": 0,
        },
        "default_val": "",
        "double_val": 0,
        "geo_val": ByteBuffer {
          "buffer": Buffer [],
          "limit": 0,
          "littleEndian": false,
          "markedOffset": -1,
          "noAssert": false,
          "offset": 0,
        },
        "int_val": Long {
          "high": 0,
          "low": 0,
          "unsigned": false,
        },
        "password_val": "",
        "str_val": "Max",
        "uid_val": Long {
          "high": 0,
          "low": 0,
          "unsigned": true,
        },
        "val": "str_val",
      },
      "predicate": "name",
      "subject": "_:tmp1",
    },
     {
      "facets": Array [],
      "label": "",
      "lang": "",
      "objectId": "_:tmp1",
      "objectType": 0,
      "objectValue": null,
      "predicate": "friend",
      "subject": "_:tmp0",
    },
  ],
}
`;

exports[`toSetMutation create nquads from object 1`] = `
 {
  "del": Array [],
  "schema": Array [],
  "set": Array [
     {
      "facets": Array [],
      "label": "",
      "lang": "",
      "objectId": "",
      "objectType": 10,
      "objectValue":  {
        "bool_val": false,
        "bytes_val": ByteBuffer {
          "buffer": Buffer [],
          "limit": 0,
          "littleEndian": false,
          "markedOffset": -1,
          "noAssert": false,
          "offset": 0,
        },
        "date_val": ByteBuffer {
          "buffer": Buffer [],
          "limit": 0,
          "littleEndian": false,
          "markedOffset": -1,
          "noAssert": false,
          "offset": 0,
        },
        "datetime_val": ByteBuffer {
          "buffer": Buffer [],
          "limit": 0,
          "littleEndian": false,
          "markedOffset": -1,
          "noAssert": false,
          "offset": 0,
        },
        "default_val": "",
        "double_val": 0,
        "geo_val": ByteBuffer {
          "buffer": Buffer [],
          "limit": 0,
          "littleEndian": false,
          "markedOffset": -1,
          "noAssert": false,
          "offset": 0,
        },
        "int_val": Long {
          "high": 0,
          "low": 0,
          "unsigned": false,
        },
        "password_val": "",
        "str_val": "Paul",
        "uid_val": Long {
          "high": 0,
          "low": 0,
          "unsigned": true,
        },
        "val": "str_val",
      },
      "predicate": "name",
      "subject": "_:tmp0",
    },
     {
      "facets": Array [],
      "label": "",
      "lang": "",
      "objectId": "",
      "objectType": 3,
      "objectValue":  {
        "bool_val": false,
        "bytes_val": ByteBuffer {
          "buffer": Buffer [],
          "limit": 0,
          "littleEndian": false,
          "markedOffset": -1,
          "noAssert": false,
          "offset": 0,
        },
        "date_val": ByteBuffer {
          "buffer": Buffer [],
          "limit": 0,
          "littleEndian": false,
          "markedOffset": -1,
          "noAssert": false,
          "offset": 0,
        },
        "datetime_val": ByteBuffer {
          "buffer": Buffer [],
          "limit": 0,
          "littleEndian": false,
          "markedOffset": -1,
          "noAssert": false,
          "offset": 0,
        },
        "default_val": "",
        "double_val": 1.85,
        "geo_val": ByteBuffer {
          "buffer": Buffer [],
          "limit": 0,
          "littleEndian": false,
          "markedOffset": -1,
          "noAssert": false,
          "offset": 0,
        },
        "int_val": Long {
          "high": 0,
          "low": 0,
          "unsigned": false,
        },
        "password_val": "",
        "str_val": "",
        "uid_val": Long {
          "high": 0,
          "low": 0,
          "unsigned": true,
        },
        "val": "double_val",
      },
      "predicate": "size",
      "subject": "_:tmp0",
    },
  ],
}
`;

exports[`toSetMutation uses _uid_ as subject 1`] = `
 {
  "del": Array [],
  "schema": Array [],
  "set": Array [
     {
      "facets": Array [],
      "label": "",
      "lang": "",
      "objectId": "",
      "objectType": 10,
      "objectValue":  {
        "bool_val": false,
        "bytes_val": ByteBuffer {
          "buffer": Buffer [],
          "limit": 0,
          "littleEndian": false,
          "markedOffset": -1,
          "noAssert": false,
          "offset": 0,
        },
        "date_val": ByteBuffer {
          "buffer": Buffer [],
          "limit": 0,
          "littleEndian": false,
          "markedOffset": -1,
          "noAssert": false,
          "offset": 0,
        },
        "datetime_val": ByteBuffer {
          "buffer": Buffer [],
          "limit": 0,
          "littleEndian": false,
          "markedOffset": -1,
          "noAssert": false,
          "offset": 0,
        },
        "default_val": "",
        "double_val": 0,
        "geo_val": ByteBuffer {
          "buffer": Buffer [],
          "limit": 0,
          "littleEndian": false,
          "markedOffset": -1,
          "noAssert": false,
          "offset": 0,
        },
        "int_val": Long {
          "high": 0,
          "low": 0,
          "unsigned": false,
        },
        "password_val": "",
        "str_val": "Paul 2",
        "uid_val": Long {
          "high": 0,
          "low": 0,
          "unsigned": true,
        },
        "val": "str_val",
      },
      "predicate": "name2",
      "subject": "0xd8450668580bf30f",
    },
  ],
}
`;
